### PR TITLE
Add `choiceController` and `choiceObserver` functions

### PR DESCRIPTION
**Pull request by akrmn**
Closes https://github.com/digital-asset/daml/issues/11362
Daml PR: https://github.com/digital-asset/daml/pull/16088
DA-GHC PR: https://github.com/digital-asset/ghc/pull/156

### CODE REVIEW

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala [hunk 1]
```diff
@@ -277,7 +277,7 @@ private[lf] class PackageInterface(signatures: PartialFunction[PackageId, Packag
     lookupInterface(ifaceName, context).flatMap(
       _.choices
         .get(chName)
-        .toRight(LookupError.NotFound(Reference.TemplateChoice(ifaceName, chName), context))
+        .toRight(LookupError.NotFound(Reference.InterfaceChoice(ifaceName, chName), context))
```
Comment (akrmn) [inline daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala:280]: tangential change, this used to give the wrong error message when the choice was not found

File: daml-lf/archive/src/main/protobuf/com/daml/daml_lf_dev/daml_lf_1.proto [hunk 2]
```diff
@@ -1009,6 +1009,23 @@ message Expr {
     Expr expr = 2;
   }
 
+  // Obtain the controllers for a choice.
+  // *Available in versions >= 1.dev*
+  message ChoiceController {
+    TypeConName template = 1;
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/archive/src/main/protobuf/com/daml/daml_lf_dev/daml_lf_1.proto:1015]: :+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala [hunk 3]
```diff
@@ -394,7 +394,11 @@ private[lf] final class Compiler(
         ).foreach(addDef)
       }
 
-      tmpl.choices.values.foreach(x => addDef(compileTemplateChoice(tmplId, tmpl, x)))
+      tmpl.choices.values.foreach(choice => {
+        addDef(compileTemplateChoice(tmplId, tmpl, choice))
+        addDef(compileChoiceController(tmplId, tmpl.param, choice))
+        addDef(compileChoiceObserver(tmplId, tmpl.param, choice))
+      })
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:401]: Why not ? 

```suggestion
      tmpl.choices.values.foreach{choice => 
        addDef(compileTemplateChoice(tmplId, tmpl, choice))
        addDef(compileChoiceController(tmplId, tmpl.param, choice))
        addDef(compileChoiceObserver(tmplId, tmpl.param, choice))
      }
```

File: daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/language/src/main/scala/com/digitalasset/daml/lf/language/PackageInterface.scala:280]: Good catch.

**Review by remyhaemmerle-da — APPROVED** [review]
Looks good to me.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala (hunk 3 above)
Comment (akrmn) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Compiler.scala:401]: thanks!

**Pull request merged**